### PR TITLE
Tailwind prefix

### LINKED ISSUE #509: [Feature]: Add tailwind `prefix` option to components.json
### Describe the feature

Current CLI doesn't support prefix option

Related [PR](https://github.com/radix-vue/shadcn-vue/pull/324) and [Discussion](https://github.com/radix-vue/shadcn-vue/discussions/444)

We welcome any help 💚 

This link might be helpful before prefix support https://github.vue.tailwind-prefix.cbass.dev/ thanks to Tobias Schenk for sharing

### Additional information

- [ ] I intend to submit a PR for this feature.
- [ ] I have already implemented and/or tested this feature.

### PR DESCRIPTION
**Pull request by zernonia**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by zernonia** [comment]
@sadeghbarati  there was some problem back then when I'm implementing the tw prefix, but I just had a look today and seems like the snapshot and everything looks good. Can you help review and check if it's all good? 😁

**Comment by sadeghbarati** [comment]
image


Just switch to your branch build the CLI and upload it to Stackblitz, seems not working yet, which is why I don't like test files 😆 

https://stackblitz.com/edit/vitejs-vite-dzqk4g

**Comment by zernonia** [comment]
Can you check again @sadeghbarati ? 😁

**Comment by sadeghbarati** [comment]
@zernonia [fix: cater for cn function](https://github.com/radix-vue/shadcn-vue/pull/619/commits/0f551c85e61a246216211033fb3788ed685f2805) did fixed the cn prefix

I see prefix applied to all `imports` in all ui `.ts` files
https://stackblitz.com/edit/vitejs-vite-dzqk4g

image

**Comment by zernonia** [comment]
Ahh nice catch @sadeghbarati ! Fixing!!!!

**Comment by sadeghbarati** [comment]
I think we should setup `pkg.pr.new` for shadcn-vue `cli` and `module` packages too,
Stackblitz upload limit will prevent me to test future PRs

**Comment by zernonia** [comment]
Yeah we can do that.. that would really help to test out the PR 😁

File: apps/www/src/lib/registry/new-york/ui/alert-dialog/AlertDialogCancel.vue [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import { type HTMLAttributes, computed } from 'vue'
 import { AlertDialogCancel, type AlertDialogCancelProps } from 'radix-vue'
 import { cn } from '@/lib/utils'
-import { buttonVariants } from '@/lib/registry/new-york/ui/button'
+import { buttonVariants } from '@/lib/registry/default/ui/button'
```
Comment (sadeghbarati) [inline apps/www/src/lib/registry/new-york/ui/alert-dialog/AlertDialogCancel.vue]:
```suggestion
import { buttonVariants } from '@/lib/registry/new-york/ui/button'
```

**Comment by zernonia** [comment]
Yup not as easy as I thought hahaha.. using`StringLiteral` is not powerful enough.. will convert to draft for now

**Comment by zernonia** [comment]
@sadeghbarati  I found this goldmind! https://vue-metamorph.dev/

It simplified the transformation soooooo much!!! I've refactor the transformer entirely to use `vue-metamorph`! Do help test it out ya!

**Comment by sadeghbarati** [comment]
I hit the upload limit 😭 

I think `vue-metamorph` should be in `dependencies`

image

---

### These tests are in TypeScript project

#### Button `index.ts`

All UI `ts` files import are still prefixed

- CVA `defaultVariants` has issue with prefixes (this is not just in the button, the component that is using CVA is having this issue)
- interface types/props also have the same issue
- `withDefaults` also has the same issue


image


#### `Button.vue`

image


---

#### `auto-form`, `chart-area|bar|donut|line`
image

**Comment by sadeghbarati** [comment]
Now I understand why Codemod Team is having hard time working on Vue Codemods

**Comment by sadeghbarati** [comment]
Tested JavaScript projects too, seems like `detypes` is not working well with `vue-metamorph`

#### Button `index.js`
image

**Comment by zernonia** [comment]
@sadeghbarati  Do you mind checking again? This time the transformation for prefix is much stricter to paritcular attributes and function only.

Also made a patch for `vue-eslint-parser` that `vue-metamorph` is using underneath. The `undefined` issue should be gone now

**Comment by sadeghbarati** [comment]
@zernonia Everything is fine just

- `/* METAMORPH_START */` comment is in the some of components after parsing 
- sonner `toast-options` classes in not having the prefixes (**ignore** defineProps error I just add the components for test)

  image

- `v-calendar` style section is not having the prefix which we could ignore duo to legacy package and new calendar implementation

**Comment by sadeghbarati** [comment]
in JavaScript projects, found a strange thing, happening in `AccordionTrigger` and `SelectItem` and maybe other components as well

image


`cn` placement is not formatted and also an additional `]` is added to the end of the cn function

**Comment by sadeghbarati** [comment]
> in JavaScript projects, found a strange thing, happening in `AccordionTrigger` and `SelectItem` and maybe other components as well
> 
> image
> 
> `cn` placement is not formatted and also an additional `]` is added to the end of the cn function

Looks like this is `detypes` bug not `vue-metamorph`

Related Discussion https://github.com/radix-vue/shadcn-vue/discussions/632

**Comment by zernonia** [comment]
Sorry that this ticket took awhile 😂. Creating upstream fix here https://github.com/Dunqing/detype/pull/3 .

After this merges, will start adapting the CLI to the changes in main [ui.shadcn](https://ui.shadcn.com/docs/changelog)

**Comment by zernonia** [comment]
@sadeghbarati  If you are free do test it out again 😁 I've publish a forked version of [detypes](https://github.com/unovue/detypes) to get around this issue early

**Comment by zernonia** [comment]
Closes https://github.com/unovue/shadcn-vue/issues/647 https://github.com/unovue/shadcn-vue/issues/772 https://github.com/unovue/shadcn-vue/issues/800 https://github.com/unovue/shadcn-vue/issues/732

**Review by sadeghbarati — APPROVED** [review]
@zernonia 

https://stackblitz.com/edit/vitejs-vite-ga6qzh?terminal=dev

LGTM

Please update all deps in `cli` package before merging and publishing new version of CLI

---

I faced two strange things while installing all the components `Carousel` and `Toast` had these errors on **first try**
also, toast had `index.ts` file instead of `index.js`, after removing and trying again these problems gone

**Carousel**

```
⠧ Installing carousel...
 ERROR  [@vue/compiler-sfc] Unresolvable type reference or unsupported built-in utility type  3:19:50 PM

anonymous.vue
4  |  import { cn } from '@/lib/utils'
5  |  
6  |  const props = withDefaults(defineProps<CarouselProps & WithClassAsProps>(), {
   |                                         ^^^^^^^^^^^^^
7  |    orientation: 'horizontal',
8  |  })

 ELIFECYCLE  Command failed with exit code 1.
```

**Toast**

```
 ERROR  [@vue/compiler-sfc] Unresolvable type reference or unsupported built-in utility type  3:23:27 PM

anonymous.vue
5  |  import { cn } from '@/lib/utils'
6  |  
7  |  const props = defineProps<ToastProps>()
   |                            ^^^^^^^^^^
8  |  
9  |  const emits = defineEmits<ToastRootEmits>()

 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by sadeghbarati** [comment]
`Sonner.vue` classes prop does not have the tailwind prefix

https://github.com/unovue/shadcn-vue/pull/619#issuecomment-2177663446

File: packages/cli/package.json [hunk 2]
```diff
@@ -52,11 +52,11 @@
   "dependencies": {
     "@babel/core": "^7.25.2",
     "@babel/parser": "^7.25.3",
+    "@unovue/detypes": "^0.8.4",
     "@vue/compiler-sfc": "^3.4",
```
Comment (sadeghbarati) [inline packages/cli/package.json]:
```suggestion
    "@vue/compiler-sfc": "^3.5",
```

**Comment by zernonia** [comment]
> `Sonner.vue` classes prop does not have the tailwind prefix
> 
> [#619 (comment)](https://github.com/unovue/shadcn-vue/pull/619#issuecomment-2177663446)

Gonna ignore this for now. Currently broken for main shadcn repo https://stackblitz.com/edit/stackblitz-starters-p5jxp5?file=components%2Fui%2Fsonner.jsx

**Comment by darkons** [comment]
After adding the Tailwind prefix to the configuration, I noticed that the prefix is misapplied to negative values.

For example in the `DialogContent` component:
https://github.com/unovue/shadcn-vue/blob/dev/apps/www/src/registry/new-york/ui/dialog/DialogContent.vue#L36

Results in `tw--translate-x-1/2 tw--translate-y-1/2` instead of `-tw-translate-x-1/2 -tw-translate-y-1/2`

https://v3.tailwindcss.com/docs/configuration#prefix

**Pull request merged**